driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2265,7 +2267,20 @@ public void run() {
                         if (!schemaInAgreement)
                             logger.warn("No schema agreement from live replicas after {} s. The schema may not be up to date on some nodes.", configuration.getProtocolOptions().getMaxSchemaAgreementWaitSeconds());
 
-                        ListenableFuture<Void> schemaReady = refreshSchema ? submitSchemaRefresh(targetType, targetKeyspace, targetName, targetSignature) : MoreFutures.VOID_SUCCESS;
+                        ListenableFuture<Void> schemaReady;
+                        if (refreshSchema) {
+                            schemaReady = submitSchemaRefresh(targetType, targetKeyspace, targetName, targetSignature);
+                            if (refreshNodeList) {
+                                schemaReady = Futures.transform(schemaReady, new AsyncFunction<Object, Void>() {
```

Comment:
Note that a call to `submitNodeListRefresh()` is already done in `ControlConnection.refreshSchema()` for other cases (notably when a schema change event is received), but here we need to wait on that future. I think the cost of enqueueing 2 node list refresh requests is negligible compared to the advantage of having the token metadata guaranteed to be up-to-date.